lib/stm32/fdcan.c

```diff
@@ -0,0 +1,891 @@
+/** @defgroup fdcan_file FDCAN peripheral API
+ * @ingroup peripheral_apis
+ * 
+ * @brief <b>libopencm3 STM32G4xx CAN</b>
+ * 
+ * @version 1.0.0
+ * 
+ * @author @htmlonly &copy; @endhtmlonly 2021 Eduard Drusa <ventyl86 at netkosice dot sk>
+ * 
+ * Devices can have up to three FDCAN peripherals residing in one FDCAN block. The peripherals 
+ * support both CAN 2.0 A and B standard and Bosch FDCAN standard. FDCAN frame format and
+ * bitrate switching is supported. The peripheral has several filters for incoming messages that
+ * can be distributed between two FIFOs and three transmit mailboxes. For transmitted messages
+ * it is possible to opt for event notification once message is transmitted.
+ * 
+ * LGPL License Terms @ref lgpl_license
+*/
+
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2021 Eduard Drusa <ventyl86 at netkosice dot sk>
+ *
+ * This library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <libopencm3/stm32/fdcan.h>
+#include <libopencm3/stm32/rcc.h>
+
+
+/* Timeout for FDCAN_CCCR register INIT bit to accept set value.
+ * This timeout is required because FDCAN uses two different clocks
+ * feeding two different portions of block. There can be slight delay
+ * based on how clocks are set up. While amount of FDCAN_clk /
+ * FDCAN_pclk combinations is high and clock speeds may vary a lot,
+ * following value has been choosen as sane default.
+ */
+#define FDCAN_CCCR_INIT_TIMEOUT		0x0000FFFF
```

Comment:
people are almost definitely going to want to tune that themselves, if you're introducing a world where functions have a timeout.  Also, please definitely consider adding api methods that allow the functions to be used in non-blocking fashion.